Abstract iptables in proxier.go

BenTheElder opened the pull request:
Abstract opening/closing/deleting all portals to a ProxyManager
interface so we can later choose between iptables and pf.

See: https://github.com/GoogleCloudPlatform/kubernetes/issues/2170; in particular: https://github.com/GoogleCloudPlatform/kubernetes/issues/2170#issuecomment-61991086

BenTheElder committed:
Abstract iptables in proxier.go

Abstract opening/closing/deleting all portals to a ProxyManager
interface so we can later choose between iptables and pf.

See: https://github.com/GoogleCloudPlatform/kubernetes/issues/2170; in particular: https://github.com/GoogleCloudPlatform/kubernetes/issues/2170#issuecomment-61991086

BenTheElder commented:
For previous conversation/PR see: https://github.com/GoogleCloudPlatform/kubernetes/pull/5961

(I cleaned up the commit history, move it to a branch and brought it up to date with master, so I've moved the PR to here form the previous one.)

BenTheElder commented:
Small note, I get some unrelated errors when building on OSX that I need to look into, so There'll be a few commits as I use travis to tell me where I missed updating field names / types. I should have access to a linux box to build on locally again in a day or two.

Sorry about that.

kelseyhightower commented on pkg/proxy/iptables_portalmanager.go:
```diff
@@ -0,0 +1,272 @@
+limitations under the License.
+*/
+
+package proxy
```
If there are no dependancies on the proxy package should we consider moving this to a new portal package?

```
pkg/proxy/portalmanager
```

Then we cloud move this under

```
pkg/proxy/portalmanager/iptables/
```

BenTheElder replied:
There's at least one small depend. 
The type proxy.ServicePortName https://github.com/BenTheElder/kubernetes/blob/abstract_iptables/pkg/proxy/loadbalancer.go#L38

Which is used a fair bit currently.

BenTheElder replied:
Perhaps ServicePortName could go in /pkg/types ?
I'm unsure or where it belongs if not in proxy.

brendandburns commented on pkg/proxy/iptables_portalmanager.go:
```diff
@@ -0,0 +1,272 @@
+	"github.com/golang/glog"
+)
+
+type IptablesPortalManager struct {
```
godoc this.

BenTheElder replied:
I'll do that shortly, I went out for a walk, beautiful day out here. :)

brendandburns commented:
Generally LGTM, but this _has_ to get run through e2e before we submit it.

BenTheElder committed:
godoc IptablesPortalManager

BenTheElder commented:
Any idea why shippable just failed?
I only added comment lines...

quinton-hoole commented:
Shippable is currently unreliable as a build signal.  But you definitely need to run the e2e tests successfully.   Please paste the results of a successful e2e run here.

BenTheElder commented:
will do. I need to install the google cloud sdk it looks like, i just have app engine for go/python sdks atm.

BenTheElder commented:
I don't want to enable billing on my google cloud account (i'm a student.) So I'm going to try to run them with vagrant. I haven't done that on my fresh fedora 21 setup yet so it might be a bit before I get that done. Sorry about that. I've just been using hack/test-go.sh thusfar.

BenTheElder commented:
First attempt: https://gist.github.com/BenTheElder/a9d6de168ef858a7fb3c

BenTheElder commented:
It seems to run fine but then after a bit it goes to start another cluster and fails. this is probably because i'm calling it via sudo, i needed to fix my docker permissions. will post back with another attempt later.

BenTheElder commented:
I've been fiddling around with running it, vagrant up / starting the e2e cluster seems to have timeout errors. I'm not very familiar with vagrant or the e2e tests but here's a paste of my latest attempt:
`sudo env "PATH=$PATH" "KUBERNETES_PROVIDER=vagrant" sh -c 'hack/e2e-test.sh > /e2eout2.txt 2>&1'` results in:
https://gist.github.com/BenTheElder/f8eb6a503175916e68a8

BenTheElder commented:
Going to do some more debug later, but i seem to be having issues with cluster/kube-up.sh
It gives me the following at the end: 
````
Waiting for each minion to be registered with cloud provider
current-context: "vagrant"
Running: ./cluster/../cluster/vagrant/../../cluster/../cluster/vagrant/../../_output/dockerized/bin/linux/amd64/kubectl --kubeconfig=/root/.kubernetes_vagrant_kubeconfig get minions -o template -t {{range.items}}{{.id}}:{{end}}
F0421 15:45:20.871726   14884 get.go:67] Client error processing command: Get https://10.245.1.2:443/api/v1beta1/minions: dial tcp 10.245.1.2:443: i/o timeout
````
Since this fails, once I start the e2e tests then partway through it goes to start up another cluster it seems and gets this error, stopping the test. I'll try to debug it later, I'm not sure what exactly is up...

BenTheElder commented:
Got a little further. I killed all of the virtualbox vms and started over.
From kubernetes dir:
````shell
su
cd <kubernetes dir>
export KUBERNETES_PROVIDER=vagrant
cluster/kube-up
hack/e2e-test.sh > ~/e2eout.txt 2>&1
````
http://paste.fedoraproject.org/214087/54093142/

errors with:
````shell
Running: /home/bentheelder/code/kubernetes/src/github.com/GoogleCloudPlatform/kubernetes/cluster/../cluster/vagrant/../../cluster/vagrant/../../_output/dockerized/bin/linux/amd64/kubectl --kubeconfig=/root/.kubernetes_vagrant_kubeconfig --match-server-version get pods --template={{range.items}}{{.currentState.status}} {{end}}
/home/bentheelder/code/kubernetes/src/github.com/GoogleCloudPlatform/kubernetes/hack/e2e-internal/e2e-status.sh: line 64: statuses[@]: unbound variable
2015/04/21 18:01:30 Error running get status: exit status 1
2015/04/21 18:01:30 Testing requested, but e2e cluster not up!
exit status 1
````

BenTheElder commented:
Had a little spare time and got vagrant setup and ran the tests again on my macbook, got the same results and eventual error:

````shell
 Vagrant doesn't need special preparations for e2e tests
F0424 21:48:36.084088   12114 command.go:442] open /Users/benjamin/.kubernetes_vagrant_kubeconfig: no such file or directory
current-context: ""
Running: /Users/benjamin/code/kubernetes/src/github.com/GoogleCloudPlatform/kubernetes/cluster/../cluster/vagrant/../../cluster/vagrant/../../_output/dockerized/bin/darwin/amd64/kubectl --kubeconfig=/Users/benjamin/.kubernetes_vagrant_kubeconfig --match-server-version version
Error: stat /Users/benjamin/.kubernetes_vagrant_kubeconfig: no such file or directory
2015/04/24 21:48:36 Error running get status: exit status 1
2015/04/24 21:48:36 Running: up
Vagrant doesn't need special preparations for e2e tests
Starting cluster using provider: vagrant
... calling verify-prereqs
... calling kube-up
Using credentials: vagrant:vagrant
Bringing machine 'master' up with 'virtualbox' provider...
Bringing machine 'minion-1' up with 'virtualbox' provider...
==> master: VirtualBox VM is already running.
==> minion-1: VirtualBox VM is already running.
2015/04/24 21:48:43 Error running up: exit status 1
2015/04/24 21:48:43 Error starting e2e cluster. Aborting.
exit status 1
````
The error seems to be unrelated to the code though, it looks like it's bugging out trying to setup the vagrant cluster even though the cluster is already up.

derekwaynecarr commented:
Right now e2e does not tear down a cluster when it's done, so it's not re-entrant 

Sent from my iPhone

> On Apr 24, 2015, at 9:56 PM, Benjamin Elder <notifications@github.com> wrote:
> 
> Had a little spare time and got vagrant setup and ran the tests again on my macbook, got the same results and eventual error:
> 
>  Vagrant doesn't need special preparations for e2e tests
> F0424 21:48:36.084088   12114 command.go:442] open /Users/benjamin/.kubernetes_vagrant_kubeconfig: no such file or directory
> current-context: ""
> Running: /Users/benjamin/code/kubernetes/src/github.com/GoogleCloudPlatform/kubernetes/cluster/../cluster/vagrant/../../cluster/vagrant/../../_output/dockerized/bin/darwin/amd64/kubectl --kubeconfig=/Users/benjamin/.kubernetes_vagrant_kubeconfig --match-server-version version
> Error: stat /Users/benjamin/.kubernetes_vagrant_kubeconfig: no such file or directory
> 2015/04/24 21:48:36 Error running get status: exit status 1
> 2015/04/24 21:48:36 Running: up
> Vagrant doesn't need special preparations for e2e tests
> Starting cluster using provider: vagrant
> ... calling verify-prereqs
> ... calling kube-up
> Using credentials: vagrant:vagrant
> Bringing machine 'master' up with 'virtualbox' provider...
> Bringing machine 'minion-1' up with 'virtualbox' provider...
> ==> master: VirtualBox VM is already running.
> ==> minion-1: VirtualBox VM is already running.
> 2015/04/24 21:48:43 Error running up: exit status 1
> 2015/04/24 21:48:43 Error starting e2e cluster. Aborting.
> exit status 1
> The error seems to be unrelated to the code though, it looks like it's bugging out trying to setup the vagrant cluster even though the cluster is already up.
> 
> —
> Reply to this email directly or view it on GitHub.
>

BenTheElder commented:
@derekwaynecarr so what is the correct way to run the tests?
Is there a workaround?
I'm just calling kube-up.sh and then e2e-test.sh with the provider set to vagrant currently.

BenTheElder commented:
Success! I'll be debugging exactly what was needed to get it working in a moment (it looks like I just needed to `export NUM_MINIONS=2` to get it working on vagrant.

I've gotten it running on my laptop both on https://github.com/GoogleCloudPlatform/kubernetes/commit/86751e8c90a3c0e852afb78d26cb6ba8cdbc37ba and then on https://github.com/BenTheElder/kubernetes/commit/064dbf98fa7fdaecd79e702d8046a12c10e3cea7

full results on  https://github.com/BenTheElder/kubernetes/commit/064dbf98fa7fdaecd79e702d8046a12c10e3cea7: https://gist.github.com/BenTheElder/e64f8805012d367759ce
Short version:
````shell
Summarizing 4 Failures:
 
[Fail] Services [It] should correctly serve identically named services in different namespaces on different external IP addresses 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:389
 
[Fail] Density [It] [Performance suite] should allow starting 30 pods per node 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:200
 
[Fail] Events [It] should be sent by kubelets and the scheduler about pods scheduling and running 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/events.go:122
 
[Fail] Services [It] should be able to create a functioning external load balancer 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:293
 
Ran 34 of 39 Specs in 1033.850 seconds
FAIL! -- 30 Passed | 4 Failed | 1 Pending | 4 Skipped F0503 21:25:02.427307    3838 driver.go:94] At least one test failed
````

-----------------------------------------

Probably not necessary but;
Full results on https://github.com/GoogleCloudPlatform/kubernetes/commit/86751e8c90a3c0e852afb78d26cb6ba8cdbc37ba: https://gist.github.com/BenTheElder/d570873ed9a5e2990d81
Short Version:

````shell
Summarizing 8 Failures:
 
[Fail] Density [It] [Performance suite] should allow starting 30 pods per node 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:136
 
[Fail] Pods [It] should contain environment variables for services 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:374
 
[Timeout...] Services [It] should serve a basic endpoint from pods 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:256
 
[Fail] ReplicationController [It] should serve a basic image on each replica with a public image 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/rc.go:137
 
[Fail] Density [It] [Performance suite] should allow starting 3 pods per node 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:136
 
[Fail] kubectl update-demo [It] should scale a replication controller 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:304
 
[Fail] Pods [It] should be restarted with a docker exec "cat /tmp/health" liveness probe 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pods.go:54
 
[Fail] kubectl update-demo [It] should create and stop a replication controller 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/util.go:304
 
Ran 37 of 42 Specs in 2198.465 seconds
FAIL! -- 29 Passed | 8 Failed | 1 Pending | 4 Skipped F0503 19:58:03.207184    5230 driver.go:94] At least one test failed
````

derekwaynecarr commented:
I have not had an opportunity to review your PR in detail, but feel free to open an issue if you see those failures on a vanilla Vagrant cluster and I will look to try and get them resolved.

BenTheElder commented:
@derekwaynecarr I'll go do that.

BenTheElder commented:
After rebasing this up to https://github.com/GoogleCloudPlatform/kubernetes/commit/1b7749b8d94ad40db1d908df28ea80cee40a24ac I have:
````shell
Summarizing 1 Failure:

[Fail] Events [It] should be sent by kubelets and the scheduler about pods scheduling and running 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/events.go:122

Ran 38 of 43 Specs in 973.040 seconds
FAIL! -- 37 Passed | 1 Failed | 1 Pending | 4 Skipped F0505 21:12:09.015275   15348 driver.go:94] At least one test failed
````
I'm not sure why this is failing; but vagrant as a provider is working smoother on fedora at least; and i'm caught up to the latest changes to proxy.

BenTheElder commented:
I've  been working on getting vagrant to run smoother, and I think I've gotten satisfactory results with e2e on vangrat, after switching to my desktop and deleting all old virtual boxes `hack/e2e-test.sh` gives:

````shell
Summarizing 1 Failure:

[Fail] Density [It] [Performance suite] should allow starting 30 pods per node 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:135

Ran 38 of 43 Specs in 698.795 seconds
FAIL! -- 37 Passed | 1 Failed | 1 Pending | 4 Skipped F0506 14:39:27.966103   12674 driver.go:94] At least one test failed
````

It looks like there are just performance issues related to running under virtualbox, no problems with the code here.

See https://github.com/GoogleCloudPlatform/kubernetes/issues/7719 for discussion about vagrant and e2e.

BenTheElder committed:
Create ProxierIPs struct for use with PortalManager

Create a ProxierIPs struct for embedding in Proxier to contain the
listenIP and hostIP fields for passing to PortalManager.

BenTheElder commented:
Woohoo! Just finished running e2e on my mac:
```shell
Ran 38 of 43 Specs in 773.538 seconds
SUCCESS! -- 38 Passed | 0 Failed | 1 Pending | 4 Skipped I0508 14:00:05.678822   30482 driver.go:96] All tests pass
````

As a side note: Forcing vagrant to make new virtual boxes for each run seems to do the trick. I'll see If i can find out why that isn't already the case and get that fixed separately.

derekwaynecarr commented:
@BenTheElder - I think https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/vagrant/util.sh#L298  needs to call kube-down a'la https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/util.sh#L846

BenTheElder commented:
@derekwaynecarr  I was actually just writing an issue for that.
The second problem is that when i call kube-down manually it misses the second minion.

derekwaynecarr commented:
Is NUM_MINIONS not exported?  I think e2e may manually be changing the value, and if they are, they need to pass that value back into the invocation of the bash shell.

BenTheElder commented:
I'm not sure. The change i wrote previously sets NUM_MINIONS in config-test.sh prior to sourcing config-default.sh which in turn exports it in theory, however if I do echo $NUM_MINIONS i get nothing.

roberthbailey commented:
How are you deleting your e2e cluster? If you run `go run hack/e2e.go -v --down` then it should pick up the variables from `config-test.sh` for your provider (see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#flag-options).

BenTheElder commented:
@roberthbailey I am doing just that but the vagrant scripts do not do anything when instructed to run test teardown (they echo "Vagrant ignores tear-down").  See https://github.com/GoogleCloudPlatform/kubernetes/issues/7978 I've just opened.

roberthbailey commented:
I just saw that one. Gotta love an "echo TODO". :)

BenTheElder commented:
Haha yeah. I didn't think anything of it at first since I'm new; but I checked aws's scripts today and they call kube-down.

I'm probably the only one using vagrant to run tests anyhow. 
It's handy to be able to test from anywhere on my laptop though so I'll see if I can get those todo's sorted.

roberthbailey commented:
> I'm probably the only one using vagrant to run tests anyhow. 

I think that a couple of RedHat folks are running e2es on vagrant as well. For example, see #6817 from @timothysc.

BenTheElder commented:
Good to know. Fix has been merged.

thockin commented:
I'm still sort of backlogged, but I will be getting to this soon.

On Fri, May 8, 2015 at 12:02 PM, Benjamin Elder <notifications@github.com>
wrote:

> Good to know. Fix has been merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7032#issuecomment-100329370>
> .
>

BenTheElder commented:
This PR is pretty broken at this point considering how many changes have been made to the proxy package since then.

Also given that rkt and docker seem to have no plans to natively run on OSX, BSD, etc anytime soon we probably won't have any use for this for quite some time, if ever. Boot2docker + native clients for OSX and Windows seem to be staying the accepted solution.

If there is demand for this, I can revisit it later.

BenTheElder commented:
Docker at least may be on FreeBSD at some point soon (https://svnweb.freebsd.org/ports?view=revision&revision=391421)  but I have no idea how much of the rest of the stack would need porting. 

However, if anyone is interested in supporting PF then ping me and after #9210 goes through I'd be happy to tinker on porting kube-proxy to PF.

ibotty commented:
it would be great to make kubernetes play well with firewalld as well, so that's another thing to consider.
